cmd/snap-update-ns/utils.go

```diff
@@ -667,3 +667,92 @@ func createWritableMimic(dir, neededBy string, sec *Secure) ([]*Change, error) {
 	}
 	return changes, nil
 }
+
+// PathIterator traverses through segments (directories and files) of some
+// path. The filesystem is never consulted, traversal is done purely in memory.
+//
+// The iterator is useful in implementing secure traversal of absolute paths
+// using the common idiom of opening the root directory followed by a chain of
+// openat calls.
+//
+// A simple example on how to use the iterator:
+// ```
+// pi := NewPathIterator(path)
+// for pi.Next() {
+//    // pi.PathSoFar() or pi.Segment()
+// }
+// ```
+type PathIterator struct {
+	path        string
+	left, right int
+}
+
+// NewPathIterator returns an iterator for traversing the given path.
+// The path is passed through filepath.Clean automatically.
+func NewPathIterator(path string) *PathIterator {
+	return &PathIterator{path: filepath.Clean(path)}
```

Comment:
I'm not sure I'm fond of automatic and silent cleanup of junk paths here. Where do these paths come from? Would it be useful to have a second variant of path iterator constructor that doesn't automatically accept any path?

Reply:
The iterator must use clean paths or it will provide incorrect results.

Reply:
Ok, I wasn't clear enough, sorry; what I meant is an iterator that doesn't clean path upfront and then throws an error if it's not correct.

Reply:
Ah so something like
```
func NewPathIterator(path string) (*PathIterator, error) {
  // if not clean then return error
}
```

I will look at how that would be in the code.

Reply:
Yes, something like that (maybe as a separate ctor).

Reply:
Done now